File: 백준/Gold/2146. 다리 만들기/다리 만들기.py
from collections import deque
import sys

input = sys.stdin.readline

dx = [-1,1,0,0]
dy = [0,0,-1,1]

queue = deque()

def check(i, j):
  queue.append((i, j))
  visited[i][j] = 0

  while(queue):
    x, y = queue.popleft()
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]

      if(0<=nx<n and 0<=ny<n):
        if(visited[nx][ny] == -1 and data[nx][ny] == 1):
          visited[nx][ny] = 0
          queue.append((nx, ny))

  # for i in range(n):
  #   print(visited[i])
  
def bfs(i, j):
  tmp = sys.maxsize
  queue.append((i, j))

  while(queue):
    x, y = queue.popleft()
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]

      if(0<=nx<n and 0<=ny<n):
        if(data[nx][ny] == 0 and visited[nx][ny] == -1):
          queue.append((nx,ny))
          visited[nx][ny] = visited[x][y] + 1

        if(data[nx][ny] == 1 and visited[nx][ny] == -1):
          tmp = min(tmp, visited[x][y])

  return tmp
  
n = int(input())
data = []

result = sys.maxsize

for _ in range(n):
  data.append(list(map(int, input().strip().split())))

for i in range(n):
  for j in range(n):
    visited = [[-1] * (n + 1) for _ in range(n + 1)]
    if(data[i][j] == 1 and visited[i][j] == -1):
      check(i, j)
      if(visited[i][j] == 0):
        tmp = bfs(i,j)
        result = min(result, tmp)

print(result)

File: 백준/Gold/2146. 다리 만들기/다리 만들기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int n;
    static int[][] data;
    static Integer[][] visited;
    static int result = Integer.MAX_VALUE;
    static Queue<Integer> queue = new LinkedList<>();  

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int nx, ny;

    static void check(int a, int b){
        queue.add(a);
        queue.add(b);
        visited[a][b] = 0;

        while (!queue.isEmpty()){
            int x = queue.poll();
            int y = queue.poll();

            for(int i = 0; i<4; i++){
                nx = x + dx[i];
                ny = y + dy[i];

                if(0<=nx && nx<n && 0<=ny && ny<n){
                    if(data[nx][ny] == 1 && visited[nx][ny] == null){
                        visited[nx][ny] = 0;
                        queue.add(nx);
                        queue.add(ny);
                    }
                }
            }
        }
    }

    static void bfs(int a, int b){
        queue.add(a);
        queue.add(b);

        while(!queue.isEmpty()){
            int x = queue.poll();
            int y = queue.poll();

            for(int i = 0; i<4; i++){
                nx = x + dx[i];
                ny = y + dy[i];

                if(0<=nx && nx<n && 0<=ny && ny<n){
                    if(data[nx][ny] == 0 && visited[nx][ny] == null){
                        visited[nx][ny] = visited[x][y] + 1;
                        queue.add(nx);
                        queue.add(ny);
                    }else if(data[nx][ny] == 1 && visited[nx][ny] == null){
                        result = Math.min(result, visited[x][y]);
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        data = new int[n][n];

        for(int i = 0; i<n; i++){
            String[] strings = br.readLine().split(" ");
            for(int j = 0; j<n; j++){
                data[i][j] = Integer.parseInt(strings[j]);
            }
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                visited = new Integer[n][n];
                if(data[i][j] == 1 && visited[i][j] == null){
                    check(i,j);
                    if(visited[i][j] == 0){
                        bfs(i,j);
                    }
                }
            }
        }

        System.out.println(result);

        br.close();
    }
}
